plugin-bsp/src/main/kotlin/org/jetbrains/plugins/bsp/impl/server/connection/DefaultBspConnection.kt

```diff
@@ -120,124 +128,117 @@ private class CancelableInvocationHandlerWithTimeout(
 }
 
 private val log = logger<DefaultBspConnection>()
-private const val CONNECT_SUBTASK_ID = "bsp-file-connection-connect"
+private const val CONNECT_TASK_ID = "bsp-autoconnect"
 
 class DefaultBspConnection(
   private val project: Project,
   private val connectionDetailsProviderExtension: ConnectionDetailsProviderExtension,
 ) : BspConnection {
   private var connectionDetails: BspConnectionDetails? = null
-
   private var server: JoinedBuildServer? = null
   private var capabilities: BazelBuildServerCapabilities? = null
-
   private var bspProcess: Process? = null
-  private var disconnectActions: MutableList<() -> Unit> = mutableListOf()
+  private val disconnectActions: MutableList<() -> Unit> = mutableListOf()
   private val timeoutHandler = TimeoutHandler { Registry.intValue("bsp.request.timeout.seconds").seconds }
-
-  override suspend fun connect(taskId: Any) {
-    try {
-      if (!isConnected()) {
-        doConnect(taskId)
-      }
-    } catch (e: Exception) {
-      bspProcess?.also {
-        if (it.isAlive) it.destroyForcibly()
+  private val mutex = Mutex()
+
+  override suspend fun connect() {
+    mutex.withLock {
```

Comment:
**suggestion (bug_risk):** Consider moving disconnect() call outside the lock to avoid potential deadlocks

The current implementation calls disconnect() while holding the lock, which could lead to deadlocks if disconnect() tries to acquire the same lock. Moving this call outside the lock would be safer.